src/expression_tree.rs

```diff
@@ -129,24 +99,290 @@ impl Expression {
     }
 
     //evaluates, then uses the approximation modifier on the tree
-    pub fn approximate<A: Modifier, E: Modifier, S: Modifier>(&self, approximator: &A, evaluator: &E, simplifier: &S) -> Result<f32, ()> {
+    pub fn approximate<A: Modifier, E: Modifier, S: Modifier>(&self, approximator: &A, evaluator: &E, simplifier: &S) -> Result<Numeric, ()> {
         let mut expr = self.clone();
 
         loop {
             expr = expr.evaluate(evaluator, simplifier);
-            if !expr.mod_iterate(approximator) {
+            if !approximator.modify(&mut expr) {
```

Comment:
Is this guaranteed to eventually be true?

Reply:
it depends. if the modifier has looping states (i.e. converting to one state, then converting back to the first), it isn't, but that would be a poorly designed modifier. we may want to bound the loop regardless.

Reply:
Yes, should be bounded, we want to avoid UB, so I think just throw a const generic in there as you said in Discord. For now that will be enough and implementor can make the decision based on environment.